snuba/datasets/errors_replacer.py

```diff
@@ -0,0 +1,427 @@
+
+def set_project_exclude_groups(project_id: int, group_ids: Sequence[int]) -> None:
+    """Add {group_id: now, ...} to the ZSET for each `group_id` to exclude,
+    remove outdated entries based on `settings.REPLACER_KEY_TTL`, and expire
+    the entire ZSET incase it's rarely touched."""
+
+    now = time.time()
+    key = get_project_exclude_groups_key(project_id)
+    p = redis_client.pipeline()
+
+    p.zadd(key, **{str(group_id): now for group_id in group_ids})
+    p.zremrangebyscore(key, -1, now - settings.REPLACER_KEY_TTL)
+    p.expire(key, int(settings.REPLACER_KEY_TTL))
+
+    p.execute()
+
+
+def get_project_needs_final_key(project_id: int) -> str:
+    return "project_needs_final:%s" % project_id
+
+
+def set_project_needs_final(project_id: int) -> Optional[bool]:
+    return redis_client.set(
+        get_project_needs_final_key(project_id), True, ex=settings.REPLACER_KEY_TTL
+    )
+
+
+def get_projects_query_flags(project_ids: Sequence[int]) -> Tuple[bool, Sequence[int]]:
+    """\
+    1. Fetch `needs_final` for each Project
+    2. Fetch groups to exclude for each Project
+    3. Trim groups to exclude ZSET for each Project
+
+    Returns (needs_final, group_ids_to_exclude)
+    """
+
+    s_project_ids = set(project_ids)
+    now = time.time()
+    p = redis_client.pipeline()
+
+    needs_final_keys = [
+        get_project_needs_final_key(project_id) for project_id in s_project_ids
+    ]
+    for needs_final_key in needs_final_keys:
+        p.get(needs_final_key)
+
+    exclude_groups_keys = [
+        get_project_exclude_groups_key(project_id) for project_id in s_project_ids
+    ]
+    for exclude_groups_key in exclude_groups_keys:
+        p.zremrangebyscore(
+            exclude_groups_key, float("-inf"), now - settings.REPLACER_KEY_TTL
+        )
+        p.zrevrangebyscore(
+            exclude_groups_key, float("inf"), now - settings.REPLACER_KEY_TTL
+        )
+
+    results = p.execute()
+
+    needs_final = any(results[: len(s_project_ids)])
+    exclude_groups = sorted(
+        {int(group_id) for group_id in sum(results[(len(s_project_ids) + 1) :: 2], [])}
+    )
+
+    return (needs_final, exclude_groups)
+
+
+class ErrorsReplacer(ReplacerProcessor):
+    def __init__(
+        self,
+        write_schema: WritableTableSchema,
+        read_schema: TableSchema,
+        required_columns: Sequence[str],
+        tag_column_map: Mapping[str, Mapping[str, str]],
+        promoted_tags: Mapping[str, Sequence[str]],
+    ) -> None:
+        super().__init__(write_schema=write_schema, read_schema=read_schema)
+        self.__required_columns = required_columns
+        self.__all_column_names = [col.escaped for col in write_schema.get_columns()]
+        self.__tag_column_map = tag_column_map
+        self.__promoted_tags = promoted_tags
+
+    def process_message(self, message: ReplacementMessage) -> Optional[Replacement]:
+        type_ = message.action_type
+        event = message.data
+
+        if type_ in (
```

Comment:
from here to the end of the file it is basically cut/paste from replacer.py.
The only difference is type annotations and the fact that the code does not depend on the dataset object anymore.